### PR TITLE
Integration Re-sync Bug Fix

### PR DESCRIPTION
**Pull request by saadmk11**
closes: https://github.com/readthedocs/readthedocs.org/issues/6123
ref: https://github.com/readthedocs/readthedocs.org/issues/6122

### CODE REVIEW

**Comment by saadmk11** [comment]
@ericholscher This is done because previously we did not sync webhooks when we create an integration but now we do so. If for any reason the webhook fails to re-sync this tries to remove the `secret` so that the integration can be used manually by the user.

**Comment by ericholscher** [comment]
Ugh, I broke something trying to push to this branch. Silly github :/

**Comment by ericholscher** [comment]
Alright, fixed up the commits, and cleaned up some of the logic. I think it makes more sense now -- I'm not 100% sold on the GitHub `setup_webhook` that will create a new webhook when we try and `sync` one we aren't managing. I don't know the best path forward on those -- perhaps we could query GitHub to get the webhook pointing at the URL?

File: readthedocs/oauth/services/github.py [hunk 1]
```diff
@@ -277,39 +280,46 @@ def update_webhook(self, project, integration):
...
-        integration.recreate_secret()
+        if not integration.secret:
+            integration.recreate_secret()
         data = self.get_webhook_data(project, integration)
         resp = None
+
+        # Handle the case where we don't have a proper provider_data set
+        # This happens with a user-managed webhook previously
+        if not integration.provider_data:
+            return self.setup_webhook(project, integration)
```
Comment (ericholscher) [inline readthedocs/oauth/services/github.py]: I think it's probably better here to actually query the webhook via GitHub: https://developer.github.com/v3/repos/hooks/#list-hooks -- we can check for the `config['url']` and then map that to our existing one. This lets us not re-create webhooks users have already created, but instead take over control of it.

File: readthedocs/oauth/services/gitlab.py [hunk 2]
```diff
@@ -418,13 +427,14 @@ def send_build_status(self, build, commit, state):
...
         """
+        resp = None
         session = self.get_session()
         project = build.project
 
         repo_id = self._get_repo_id(project)
 
         if repo_id is None:
-            return (False, None)
+            return (False, resp)
```
Comment (saadmk11) [inline readthedocs/oauth/services/gitlab.py]: Why are we returning response here? `resp` is not defined yet

File: readthedocs/oauth/services/gitlab.py [hunk 3]
```diff
@@ -443,8 +453,6 @@ def send_build_status(self, build, commit, state):
         }
         url = self.adapter.provider_base_url
 
-        resp = None
```
Comment (saadmk11) [inline readthedocs/oauth/services/gitlab.py]: Should we be removing this?

File: readthedocs/oauth/services/gitlab.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/oauth/services/gitlab.py]: Oh! Got it you defined it before. My bad :)

File: readthedocs/oauth/services/gitlab.py (hunk 3 above)
Comment (saadmk11) [reply readthedocs/oauth/services/gitlab.py]: Redefined before

File: readthedocs/oauth/services/github.py [hunk 4]
```diff
@@ -203,7 +203,7 @@ def setup_webhook(self, project, integration=None):
         """
         session = self.get_session()
         owner, repo = build_utils.get_github_username_repo(url=project.repo)
-        if integration:
+        if integration and not integration.secret:
```
Comment (saadmk11) [inline readthedocs/oauth/services/github.py]: @ericholscher When we pass the integration from `update_webhook` to `setup_webhook` on `404`the integration already has a secret so this condition always returns `False`  and creates a new integration. thats why some tests are failing.

File: readthedocs/oauth/services/base.py [hunk 5]
```diff
@@ -226,6 +226,19 @@ def get_paginated_results(self, response):
         """
         raise NotImplementedError
 
+    def get_provider_data(self, project, integration):
+        """
+        Gets provider data from GitHub Webhooks API.
```
Comment (ericholscher) [inline readthedocs/oauth/services/base.py]: This shouldn't mention GitHub, as it's on the base class.

**Comment by ericholscher** [comment]
@stsewd Would love your take on this -- I think it's a better approach, but could use a bit more review.

**Comment by stsewd** [comment]
@ericholscher I'll try to review this today

**Comment by saadmk11** [comment]
@stsewd Thanks for the update looks good :)

**Comment by ericholscher** [comment]
> If we go for not update the secret on resync, we need to update our docs https://docs.readthedocs.io/en/stable/webhooks.html#payload-validation

Yea, we should update the doc in this PR then :+1:

**Comment by stsewd** [comment]
Done, feel free to merge :)

**Pull request merged**